### PR TITLE
[WIP] Fix cpp grad accessor API

### PR DESCRIPTION
**Pull request by albanD**
Update the API to access grad in cpp to avoid unexpected thread safety issues.
In particular, with the current API, a check like `t.grad().defined()` is not thread safe.

- This introduces `t.mutable_grad()` that should be used when getting a mutable version of the saved gradient. This function is **not** thread safe.
- The `Tensor& grad()` API is now removed. We could not do a deprecation cycle as most of our call side use non-const Tensors that use the non-const overload. This would lead to most calls hitting the warning. This would be too verbose for all the users.

### CODE REVIEW

File: test/cpp/api/nn_utils.cpp [hunk 1]
```diff
@@ -41,7 +41,7 @@ TEST_F(NNUtilsTest, ClipGradNorm) {
     for (int i = 0; i < grads.size(); i++) {
       auto param = l->parameters()[i];
       auto grad = grads[i];
-      p_scale.push_back(param.grad().data().div(grad).view(-1));
+      p_scale.push_back(param.mutable_grad().data().div(grad).view(-1));
```
Comment (ezyang) [inline test/cpp/api/nn_utils.cpp:44]: So this one is a read-only use of grad right?

File: test/cpp/api/nn_utils.cpp [hunk 2]
```diff
@@ -79,7 +79,7 @@ TEST_F(NNUtilsTest, ClipGradNorm) {
   };
   for (auto norm_type : norm_types) {
     for (int i = 0; i < grads.size(); i++) {
-      l->parameters()[i].grad().data().copy_(grads[i]);
+      l->parameters()[i].mutable_grad().data().copy_(grads[i]);
```
Comment (ezyang) [inline test/cpp/api/nn_utils.cpp:82]: This one too.

Comment (albanD) [reply test/cpp/api/nn_utils.cpp:82]: Ho the .data() removed the const-ness :'( good to know that it is as bad in cpp as it is in python haha

File: test/cpp/api/nn_utils.cpp (hunk 1 above)
Comment (albanD) [reply test/cpp/api/nn_utils.cpp:44]: Yes my bad.

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 3]
```diff
@@ -33,7 +33,7 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
   if (!variable.requires_grad())
     return {};
 
-  at::Tensor& grad = variable.grad();
+  at::Tensor& grad = variable.mutable_grad();
```
Comment (ezyang) [inline torch/csrc/autograd/functions/accumulate_grad.cpp:36]: This looks wrong, we should call `mutable_grad` after we have taken out the lock

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 4]
```diff
@@ -938,7 +938,7 @@ void Reducer::runGradCallbackForVariable(
     GradCallback&& cb) {
   auto context_ptr = rpc_context_.context_ptr.load();
   if (context_ptr == nullptr) {
-    cb(variable.grad());
+    cb(variable.mutable_grad());
```
Comment (ezyang) [inline torch/csrc/distributed/c10d/reducer.cpp:941]: This one also looks very suspicious!

Comment (albanD) [reply torch/csrc/distributed/c10d/reducer.cpp:941]: All the calls in the reducer are protected by a single mutex at the top level so this should be fine.

**Review by ezyang — APPROVED** [review]

**Comment by albanD** [comment]
Even though the code compiles just fine right now (and all tests pass fine), if I add back the bad overload (for BC reasons with a warning), the compiler actually use the bad overload...
Do we actually want that? Or are we happy with a hard BC breaking here (not adding back the bad overload with a warning and merging as-is)?

cc @glaringlee

**Comment by glaringlee** [comment]
@albanD 
BC breaking is not the end of world, but it would be a bad user experience if their code suddenly won't compile. 
I am not sure how bad this will be, given all the pytorch internal code is already replaced with mutable_grad(), can we do the following?

```
Tensor& grad() {
    TORCH_WARN_ONCE("..."); // say something like non-const grad() will be deprecated in next release, call mutable_grad() instead if tensor is actually mutable, or make it const"
    return impl_->mutable_grad();
}
```

**Comment by albanD** [comment]
I did that on my local commit. And when I recompile and run the test, the warning actually fires a lot of times (when I remove the once). Note that the exact same code with just that function removed works fine!

My understanding is that the compiler uses the "non-const" overload even though it could be using the const one. And that means that potentially all users are going to see this warning if we add this overload back...
